src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const states = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+      default:
+        break;
+    }
 
+    states.push(newState);
+    currentState = newState;
```

Comment:
You are directly assigning `newState` to `currentState`, which is not necessary. Instead, you can push the cloned state into `states` array and refer to the last element of that array when you need to use the current state for the next action.

Reply:
agree with AI